src/components/questionary/QuestionaryComponentRegistry.tsx

```diff
@@ -0,0 +1,111 @@
+import React, { FunctionComponent } from 'react';
+
+import { BasicComponentProps } from 'components/proposal/IBasicComponentProps';
+import {
+  Answer,
+  DataType,
+  Question,
+  QuestionTemplateRelation,
+  Template,
+} from 'generated/sdk';
+import { Event } from 'models/QuestionaryEditorModel';
+
+import { booleanDefinition } from './questionaryComponents/Boolean/BooleanDefinition';
+import { dateDefinition } from './questionaryComponents/DatePicker/DatePickerDefinition';
+import { embellishmentDefinition } from './questionaryComponents/Embellishment/EmbellishmentDefinition';
+import { fileUploadDefinition } from './questionaryComponents/FileUpload/FileUploadDefinition';
+import { multipleChoiceDefinition as multiChoiceDefinition } from './questionaryComponents/MultipleChoice/MultipleChoiceDefinition';
+import { proposalBasisDefinition } from './questionaryComponents/ProposalBasis/ProposalBasisDefinition';
+import { sampleBasisDefinition } from './questionaryComponents/SampleBasis/SampleBasisDefinition';
+import { sampleDeclarationDefinition } from './questionaryComponents/SampleDeclaration/SampleDeclaratonDefinition';
+import { textInputDefinition } from './questionaryComponents/TextInput/TextInputDefinition';
+
+export interface FormProps<ValueObjectType> {
+  field: ValueObjectType;
+  template: Template;
+  dispatch: React.Dispatch<Event>;
+  closeMe: () => any;
+}
+
+export type FormComponent<ValueObjectType> = FunctionComponent<
+  FormProps<ValueObjectType>
+>;
+
+export interface QuestionaryComponentDefinition {
+  dataType: DataType;
+  name: string;
+  questionTemplateRelationForm: () => FormComponent<QuestionTemplateRelation>;
+  questionForm: () => FormComponent<Question>;
+  questionaryComponent: (props: BasicComponentProps) => JSX.Element | null;
+  answerRenderer: (props: { answer: Answer }) => JSX.Element | null;
+  createYupValidationSchema: ((field: Answer) => object) | null;
+  readonly: boolean; // if true then no answer will be produced
+  creatable: boolean; // if true then the question can be added to a questionary
+  icon: JSX.Element;
+}
+
+const registry = [
+  booleanDefinition,
+  dateDefinition,
+  embellishmentDefinition,
+  fileUploadDefinition,
+  multiChoiceDefinition,
+  textInputDefinition,
+  sampleDeclarationDefinition,
+  proposalBasisDefinition,
+  sampleBasisDefinition,
+];
+
+const componentMap = new Map<DataType, QuestionaryComponentDefinition>();
+registry.forEach(definition =>
+  componentMap.set(definition.dataType, definition)
+);
+export function getQuestionaryComponentDefinition(id: DataType) {
+  const definition = componentMap.get(id);
+  if (!definition) {
+    throw new Error(`Definition for ${id} was not found`);
+  }
+
+  return definition;
+}
+
+// Convenience methods below
+export const getQuestionaryComponentDefinitions = () => registry.slice(0);
```

Comment:
Is there a reason for creating a shallow copy? The filter/map in `src/components/template/QuestionPicker.tsx` already creates a new array. If the goal is to have an array that can be freely altered, we should make a deep copy. If we want to have access to the list which can be freely filtered, used for mapping, using `ReadOnly` type will make it less error-prone I think.

Reply:
Good idea, also froze the object and did the same for the backend.

Reply:
The `immer` package has a `Immutable<T>`  type, in the future may be worth using as it is usable for deep marking [like this](https://www.typescriptlang.org/play?ts=3.9.7#code/PQKhCgAIUhhB7AtgUwM6QGYCcmQJKIpZQjDgAuAngA7KQCC5SAlgMYDyARgFbKvmQAvOACQAH0gAxAK4A7fs3izREgLIBDagB51sygBpIuygD4VkAOrJ1Aaw3bjh42fGQAysnI69LiVdseXs7mAAo4iMyoyN6m5gAi6uTI5gBKyADmAKIAHtTmAELw8AA21squAHLSiJzIxK5u5FjMsung4Mi58FgCVLT4hNLk6pylWgAqJkKQ45CdSbIAJuiMLBw8fOSiAPwzogBcM3PZC8uQ8Bv8O5AA3ljWi0rFlJAA2gDSkC2QNsiU8BgZgBdQ4ERBDEZjcYfIEmAC+Bz2wGRKNRaPRGIxInarCUqAE6lQi2mrygkHJAEYAEwAZn0ZPJN0Ji0O1JpcPp5PJAHJmdz6UCANw4vG9NACQSQAAUAEpQYNhqNon1kACjESpoIpszheBcbJ8ZB7oDJUl8bLdcbXgAGIHTNnCq0UoEAOmZ9tpjuQGFeVLtkt5RO5gqAA) so it needs less manual typing.